torch/distributed/tensor/_redistribute.py

```diff
@@ -28,95 +37,538 @@ class _TransformInfo(NamedTuple):
+                for tensor_dim, mesh_dims in enumerate(self.tensor_dim_to_mesh_dim):
+                    if len(mesh_dims) > 0:
+                        out_str += f"S({tensor_dim})"
+                        out_str += f"[{', '.join([str(m) for m in mesh_dims])}]"
+                # in addition, add the partial placement
+                partial_to_mesh_dim: dict[Partial, list[int]] = {}
+                for mesh_dim, p in enumerate(self.placements):
+                    if isinstance(p, Partial):
+                        if p not in partial_to_mesh_dim:
+                            partial_to_mesh_dim[p] = []
+                        partial_to_mesh_dim[p].append(mesh_dim)
+                for p, mesh_dims in partial_to_mesh_dim.items():
+                    out_str += f"P({p.reduce_op})"
+                    out_str += f"[{', '.join([str(m) for m in mesh_dims])}]"
+            else:
+                # native dtensor style sharding representation: map from mesh
+                # dim to tensor dim
+                for mesh_dim, placement in enumerate(self.placements):
+                    if isinstance(placement, Replicate):
+                        out_str += "R"
+                    elif isinstance(placement, Shard):
+                        assert mesh_dim in self.tensor_dim_to_mesh_dim[placement.dim]
+                        out_str += f"S({placement.dim})[{self.tensor_dim_to_mesh_dim[placement.dim].index(mesh_dim)}]"
+                    else:
+                        assert isinstance(placement, Partial)
+                        out_str += f"P({placement.reduce_op})"
+            return out_str
+
+        def __repr__(self):
+            return self.__str__()
+
+        def __post_init__(self):
+            assert len(self.placements) == len(self.tensor_dim_to_mesh_dim)
+            # precompute hash after all attributes are set
+            object.__setattr__(
+                self,
+                "_hash",
+                self._compute_hash(),
+            )
+
+        def __hash__(self) -> int:
+            return self._hash if self._hash is not None else self._compute_hash()
+
+        def _compute_hash(self) -> int:
+            return hash(
+                (
+                    self.placements,
+                    self.tensor_dim_to_mesh_dim,
+                )
+            )
+
+        def __eq__(self, other: object) -> bool:
+            if not isinstance(other, DTensorRedistributePlanner.DistState):
+                return False
+            if self._hash != other._hash:
+                return False
+            return (
+                self.placements,
+                self.tensor_dim_to_mesh_dim,
+            ) == (
+                other.placements,
+                other.tensor_dim_to_mesh_dim,
+            )
+
+    @classmethod
+    def _create_cache_key(cls, device_mesh, tensor_dimension):
+        return (id(device_mesh), tensor_dimension)
+
+    def __new__(cls, device_mesh, tensor_dimension):
```

Comment:
🤔  If you're already defining `__new__` here, why do you need to also have a cached factory function that constructs this

Reply:
For a transparent caching mechanism like this, it is better to go through a function; people don't expect Foo() to potentially return a pre-existing instance of a function.